cachito/workers/pkg_managers/ruby.py

```diff
@@ -0,0 +1,69 @@
+from pathlib import Path
+
+import packagedcode.gemfile_lock as gemlock_parser
+
+from cachito.errors import ValidationError
+
+
+class GemMetadata:
+    """Gem metadata."""
+
+    def __init__(self, name, version, gem_type, path):
+        """Fill GemMetadata with data."""
+        self.name = name
+        self.version = version
+        self.type = gem_type
+        self.path = path
+
+    def __eq__(self, other):
+        if isinstance(other, GemMetadata):
+            return (
+                self.name == other.name
+                and self.version == other.version
+                and self.type == other.type
+                and self.path == other.path
+            )
+        return False
+
+
+def parse_gemlock(gemlock_path: Path):
+    """Parse dependencies from Gemfile.lock.
+
+    :param Path gemlock_path: the full path to Gemfile.lock
+    :return: list of Gems
+    """
+    dt = gemlock_parser.GemfileLockParser(str(gemlock_path)).dependency_tree
+    dependencies = []
+
+    for key in dt:
+        gem = dt[key]
+        validate_gem_metadata(gem, gemlock_path.parents[0])
+        dependencies.append(GemMetadata(gem.name, gem.version, gem.type, gem.remote))
+
+    return dependencies
+
+
+def validate_gem_metadata(gem, gemlock_dir):
```

Comment:
I'm missing a few validations here (based on the design doc):
> NOTE: Platform-dependent Gem references are not considered valid, since they can contain pre-compiled binaries.

> For any dependencies referenced from [other sources](https://docs.google.com/document/d/14KaPuBX4BtK3EvTzyJpV-TZ8uhCa7EM7FZARaLC-ZtQ/edit#bookmark=id.vsrcb15gznd), Cachito will check that each reference is to a pinned version or commit hash

And I would be careful about "pinned version", git tags are not immutable. I believe pip, NPM and Yarn support only commit hashes in Cachito, if acceptable, ruby should do the same

Reply:
Added platform check - the way bundler works with platforms has changed over the years. See how it [works](https://prathamesh.tech/2021/04/18/bundler-2-2-3-and-deployment-of-ruby-apps/) [now](https://www.moncefbelyamani.com/understanding-the-gemfile-lock-file/#platforms) x how [it used to work](https://github.com/rubygems/rubygems/issues/4269#issuecomment-758564690). TLDR: One Gemfile.lock can contain dependencies for multiple platforms and the way they are be resolved at deploy/install time changed with bundler version 2.2.0 and 2.2.3.

I decided to implement it the simplest way - support only Gemfile.lock with only one platform specified - ruby. This way, we will be able to support all bundler versions and it will be simpler to parse the resulting file (scancode-toolkit isn't able to parse files with multiple platform versions of one package).

Added check for Git reference (40 character hash). For GEM dependencies, semantic versioning is recommended, [but not enforced](https://guides.rubygems.org/patterns/), therefore the only thing that's validated is if the version isn't None (similarly for PATH).

Reply:
Just for my understanding - semantic versioning is not required, but the lockfile _does_ pin dependencies to one version, right? Maybe the version doesn't follow semver conventions, but it's still a specific pinned version?